src/async_fuse/memfs/s3_metadata.rs

```diff
@@ -543,152 +559,202 @@ impl<S: S3BackEnd + Sync + Send + 'static> MetaData for S3MetaData<S> {
-            old_node.set_name(new_name);
-            new_node.set_name(old_name);
-
-            txn.set(
-                &KeyType::INum2Node(old_ino),
-                &ValueType::Node(old_node.into_serial_node()),
-            );
-            txn.set(
-                &KeyType::INum2Node(new_ino),
-                &ValueType::Node(new_node.into_serial_node()),
-            );
+            let new_entry = {
+                let raw_node = new_parent_node.lock().await;
+                raw_node.get_entry(new_name)
+            };
 
-            if let Some(mut new_parent_node) = new_parent_node {
-                // The two parent node is not the same node
-                let old_entry = old_parent_node.get_dir_data_mut().remove(old_name).unwrap_or_else(||unreachable!("Impossible case when exchange,\
-                                                                                                                                    as {old_name} under {old_parent} is checked to be existed."));
-                let new_entry = new_parent_node.get_dir_data_mut().remove(new_name).unwrap_or_else(||unreachable!("Impossible case when exchange,\
-                                                                                                                                    as {new_name} under {new_parent} is checked to be existed."));
-                old_parent_node.insert_entry_for_rename(DirEntry::new(
-                    old_name.into(),
-                    Arc::clone(new_entry.file_attr_arc_ref()),
-                ));
-                new_parent_node.insert_entry_for_rename(DirEntry::new(
-                    new_name.into(),
-                    Arc::clone(old_entry.file_attr_arc_ref()),
-                ));
+            match new_entry {
+                None => {
+                    // new_name does not exist under new_parent
+                    if exchange {
+                        // exchange is true, new name must exist
+                        return build_enoent(
+                            new_name,
+                            new_parent,
+                            new_parent_node.lock().await.get_name(),
+                        );
+                    }
+                    // exchange is false, so we can do rename directly
+                    if old_parent == new_parent {
+                        let mut raw_node = old_parent_node.lock().await;
+                        raw_node.get_dir_data_mut().remove(old_name);
+                        raw_node.insert_entry_for_rename(DirEntry::new(
+                            new_name.into(),
+                            Arc::clone(old_entry.file_attr_arc_ref()),
+                        ));
+                    } else {
+                        let mut old_raw_node = old_parent_node.lock().await;
+                        let mut new_raw_node = new_parent_node.lock().await;
+                        old_raw_node.get_dir_data_mut().remove(old_name);
```

Comment:
Call `old_raw_node.remove_entry_for_rename` here, because it updates the `mtime` and `ctime` of the old parent.